### PR TITLE
gh-125916: Adapt functools.reduce() to Argument Clinic

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-125916

### CODE REVIEW

**Comment by skirpichev** [comment]
This coming from https://github.com/python/cpython/pull/125917.  Red alert: to make AC happy, docstring was adjusted.

I did some benchmarks.

<details>

```py
# a.py
import pyperf
from functools import reduce

f = lambda x, y: x + y
lst = list(range(5))
init = 123

runner = pyperf.Runner()
runner.bench_func('reduce(f, lst)', reduce, f, lst)
runner.bench_func('reduce(f, lst, init)', reduce, f, lst, init)
```

Run e.g. with: ``python a.py -q -o ref.json``.

</details>

with results:
| Benchmark            | ref     | patch                 |
|----------------------|:-------:|:---------------------:|
| reduce(f, lst)       | 1.27 us | 1.19 us: 1.07x faster |
| reduce(f, lst, init) | 1.43 us | 1.36 us: 1.05x faster |
| Geometric mean       | (ref)   | 1.06x faster          |

So, at least AC-backed argument processing offer some speedup.

**Comment by erlend-aasland** [comment]
Thanks Sergey, I'll review this within today. Ping me tomorrow if I forget.

File: Modules/_functoolsmodule.c [hunk 1]
```diff
@@ -932,15 +932,31 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
...
 /* reduce (used to be a builtin) ********************************************/
 
-// Not converted to argument clinic, because of `args` in-place modification.
-// AC will affect performance.
+/*[clinic input]
+_functools.reduce
+
+    function as func: object
+    iterable as seq: object
+    initial as result: object(c_default="NULL") = _functools._initial_missing
```
Comment (JelleZijlstra) [inline Modules/_functoolsmodule.c]: Do we really need to add this `_initial_missing` object?

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: I see no other options, just as in the pure-Python version.  What you could suggest instead?  ``None`` and ``...`` were rejected.

If this is no-go - perhaps reduce() shouldn't be converted to AC, until it get support for multiple signatures.

Comment (JelleZijlstra) [reply Modules/_functoolsmodule.c]: An alternative is to write the c_default as NULL and manually write an `@text_signature`. This is what is done by `_collections.deque.index`.

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: Ah, I see, that's an option.  The drawback is that the inspect.signature() can't parse this and pure-Python version will have a different signature.

I'll postpone this change till @erlend-aasland review.

Comment (serhiy-storchaka) [inline Modules/_functoolsmodule.c]: Why not simply use NULL?
```suggestion
    initial as result: object = NULL
```

Comment (erlend-aasland) [reply Modules/_functoolsmodule.c]: Yes, my gut feel when I first browsed this PR earlier today, was also to use `@text_signature`. I think we should consider it. I'm not sure how relevant differences between the C extension module and the Python version of `functools` are.

File: Modules/_functoolsmodule.c [hunk 2]
```diff
@@ -932,15 +932,31 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
...
+    /
+
+Apply a function of two arguments cumulatively to an iterable, from left to right.
+
+This efficiently reduces the iterable to a single value.  If initial is present,
+it is placed before the items of the iterable in the calculation, and serves as
+a default when the iterable is empty.
+
+For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5])
+calculates ((((1 + 2) + 3) + 4) + 5).
```
Comment (erlend-aasland) [inline Modules/_functoolsmodule.c]: Let's keep the docstring as it was. If it needs amendment, we should do so in a separate PR, because _that_ PR would then be backported to all active branches (though, of course backported manually bco. the Argument Clinic changes).

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: Unfortunately, AC requires a summary line.  The reduce() docstring doesn't fit in that format as-is.

File: Modules/_functoolsmodule.c [hunk 3]
```diff
@@ -932,15 +932,32 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
 
 /* reduce (used to be a builtin) ********************************************/
 
-// Not converted to argument clinic, because of `args` in-place modification.
-// AC will affect performance.
+/*[clinic input]
+@text_signature "($module, function, iterable[, initial], /)"
```
Comment (JelleZijlstra) [inline Modules/_functoolsmodule.c]: With Serhyi's suggestion we don't actually need `@text_signature` any more. If you remove it, you get:

```
>>> functools.reduce.__text_signature__
'($module, function, iterable, initial=<unrepresentable>, /)'
```

There is an open issue #87233 about supporting such signatures in inspect. I suggest we remove this line for now so that the signature becomes representable once that issue is fixed.
```suggestion
```

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: I'm -0 on this - lets reviewers decide if this suggestion should be applied.

To me, using ``@text_signature`` basically means: FIXME, we actually can't represent signature for this function.  I think that in most (all?) cases fixing this will require support for multiple signatures.  So, the clinic input must be manually modified.  And ``@text_signature`` marks are easily greppable.

File: Modules/_functoolsmodule.c (hunk 2 above)
Comment (erlend-aasland) [reply Modules/_functoolsmodule.c]: Right, so let's amend the docstring first (in a separate PR).

Comment (JelleZijlstra) [reply Modules/_functoolsmodule.c]: The docstring change here seems small enough that it's fine to do it also in this PR.

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: FYI: https://github.com/python/cpython/pull/126045

File: Modules/_functoolsmodule.c [hunk 4]
```diff
@@ -932,15 +932,32 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
...
+_functools.reduce
+
+    function as func: object
+    iterable as seq: object
+    initial as result: object = NULL
+    /
+
+Apply a function of two arguments cumulatively to an iterable, from left to right.
+
+This efficiently reduces the iterable to a single value.  If initial is present,
```
Comment (skirpichev) [inline Modules/_functoolsmodule.c]:
```suggestion
This effectively reduces the iterable to a single value.  If initial is present,
```

File: Modules/_functoolsmodule.c (hunk 2 above)
Comment (skirpichev) [reply Modules/_functoolsmodule.c]: pr updated to include docstrings changes.

File: Modules/_functoolsmodule.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_functoolsmodule.c]: I'm fine with both Serhiy's suggestion and `@text_signature`. Since both Jelle and Alex is in favour of Serhiy's suggestion, let's go for that.

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: Alex?

Comment (erlend-aasland) [reply Modules/_functoolsmodule.c]: > Alex?

Yes, he reacted with a thumbs-up to Jelle's https://github.com/python/cpython/pull/125999#discussion_r1817966240

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
I'll give Jelle and Serhiy a chance to chime in again before landing this.

**Comment by erlend-aasland** [comment]
I plan to land this later today (post noon ET).

**Review by JelleZijlstra — APPROVED** [review]

File: Modules/_functoolsmodule.c [hunk 5]
```diff
@@ -933,28 +933,27 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
...
+Apply a function of two arguments cumulatively to the items of an iterable, from left to right.
 
-This efficiently reduces the iterable to a single value.  If initial is present,
+This effectively reduces the iterable to a single value.  If initial is present,
 it is placed before the items of the iterable in the calculation, and serves as
 a default when the iterable is empty.
 
 For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5])
-calculates ((((1 + 2) + 3) + 4) + 5).
+calculates ((((1 + 2) + 3) + 4) + 5).")
```
Comment (serhiy-storchaka) [inline Modules/_functoolsmodule.c]:
```suggestion
calculates ((((1 + 2) + 3) + 4) + 5).
```

**Comment by erlend-aasland** [comment]
Thanks for the PR, Sergey; thanks for the reviews, Jelle and Serhiy.
